x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationMetrics.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationResult;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+
+/**
+ * Provides a common way of registering and collecting various authentication metrics.
+ *
+ * @param <C> The type of context object which is used to attach additional attributes to collected metrics.
+ */
+final class AuthenticationMetrics<C> {
+
+    public static final String METRIC_SUCCESS_COUNT = "es.security.authc.%s.success.count";
+    public static final String METRIC_FAILURES_COUNT = "es.security.authc.%s.failures.count";
```

Comment:
> Does authc stand for authenticator chain?

The `authc` is just a shorthand for authentication and `authz` for authorization. I tried to keep the same naming conventions as we have for settings.

> So naming it AuthenticatorMetrics and consider moving the wrapper to the Authenticator level.

I like the `AuthenticatorMetrics` name better. Thanks for the suggestion.

re: moving it to the authenticator level - we synced and agreed that the current approach does carry a small complexity in regards applicability to token and service account authenticators, but they are non-blocking as we can solve them. The biggest concern is that by moving the "wrapping" to the `Authenticator` level it would not be possible to collect per-realm metrics, which we want to do. The agreement is to proceed with the current listener wrapping approach and iterate later to a better fit if/when we start adding more metrics.